### PR TITLE
Add Clifford Tableau decomposition function 

### PR DESCRIPTION
**Pull request by ybc1991**
Decompose clifford tableau  into one/two qubits operations (X, Z, S, H, CNOT, and SWAP). The method is based on [Aaronson & Gottesman's paper](https://arxiv.org/abs/quant-ph/0406196) [WIP for #3639].

Issue #4182 is discovered in this PR's test.

### CODE REVIEW

File: cirq-core/cirq/optimizers/clifford_decomposition_test.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        cirq.S(qubits[0]),
+        cirq.X(qubits[1]),
+    )
+
+    ops = cirq.decompose_clifford_tableau_to_operations(qubits, args.tableau)
+    circ = cirq.Circuit(ops)
+    assert_allclose_up_to_global_phase(cirq.unitary(expect_circ), cirq.unitary(circ), atol=1e-7)
+
+
+def test_clifford_decompose_by_unitary():
```
Comment (balopat) [inline cirq-core/cirq/optimizers/clifford_decomposition_test.py:115]: I'm wondering if we can reuse the implementation of `assert_implements_consistent_protocols` in `cirq.testing.consistent_protocols`? It does have a decomposition consistent with unitary assertion...

File: cirq-core/cirq/optimizers/clifford_decomposition.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+import functools
+
+import numpy as np
+from cirq import ops, protocols, qis, sim
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def _X(q, args, operations, qubits):
```
Comment (balopat) [inline cirq-core/cirq/optimizers/clifford_decomposition.py:27]: type and return annotations are missing in all these private functions.

File: cirq-core/cirq/optimizers/clifford_decomposition.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    # [1] S. Aaronson, D. Gottesman, *Improved Simulation of Stabilizer Circuits*,
+    #     Phys. Rev. A 70, 052328 (2004). https://arxiv.org/abs/quant-ph/0406196
+    # with modification by doing it row-by-row instead.
+
+    # Suppose we have a Clifford Tableau:
+    #                   Xs  Zs
+    # Destabilizers:  [ A | B ]
+    # Stabilizers:    [ C | D ]
+    for i in range(t.n):
+        # Step 1a: Make the diagonal element of A as 1 by Hadamard gate if necessary.
```
Comment (balopat) [inline cirq-core/cirq/optimizers/clifford_decomposition.py:107]:
```suggestion
        # Step 1a: Make the diagonal element of A equal to 1 by Hadamard gate if necessary.
```

File: cirq-core/cirq/optimizers/clifford_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        if not t.xs[i, i] and t.zs[i, i]:
+            _H_with_ops(i)
+        # Step 1b: Make the diagonal element of A as 1 by swapping gate if necessary.
```
Comment (balopat) [inline cirq-core/cirq/optimizers/clifford_decomposition.py:110]:
```suggestion
        # Step 1b: Make the diagonal element of A equal to 1 by SWAP gate if necessary.
```

**Review by balopat — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/optimizers/clifford_decomposition.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+def _Z(q, args, operations, qubits):
+    args.axes = [q]
+    protocols.act_on(ops.Z, args, allow_decompose=False)
+    operations.append(ops.Z(qubits[q]))
+
+
+def _Sdg(q, args, operations, qubits):
+    # Apply the tableau with S^+, so the inverse of operation is S
+    args.axes = [q]
+    protocols.act_on(ops.ZPowGate() ** 1.5, args, allow_decompose=False)
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply cirq-core/cirq/optimizers/clifford_decomposition.py:42]: Yes, it is easier to read.

File: cirq-core/cirq/optimizers/clifford_decomposition.py (hunk 2 above)
Comment (ybc1991) [reply cirq-core/cirq/optimizers/clifford_decomposition.py:27]: Done

File: cirq-core/cirq/optimizers/clifford_decomposition_test.py (hunk 1 above)
Comment (ybc1991) [reply cirq-core/cirq/optimizers/clifford_decomposition_test.py:115]: After a brief look of this function, I think it is more suitable to put into the `n_qubit_clifford_gate` class, which is the next step when all these Clifford Tableau helper functions are done. Wdyt?

Comment (balopat) [reply cirq-core/cirq/optimizers/clifford_decomposition_test.py:115]: I guess that can work - and will be consistent with having `Gate`s having a decomposition - but on the other hand there is nothing wrong with having it on the tableau and then the `Gate` delegating to the tableau's method, so I think I'm okay to have it here, as this way the Aaaronson-Gottesman paper implementations are kept closer together.

File: cirq-core/cirq/ops/common_gates.py [hunk 6]
```diff
@@ -1073,8 +1073,11 @@ def _act_on_(self, args: Any):
                     tableau.zs[:, q2].copy(),
                     tableau.xs[:, q2].copy(),
                 )
+                tableau.rs[:] ^= tableau.xs[:, q2] & tableau.zs[:, q2]
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply cirq-core/cirq/ops/common_gates.py]: Done. And I updated for the new act_on style as well. PTAL

**Comment by 95-martin-orion** [comment]
This should be reassigned from @balopat to one of the remaining Google Cirq maintainers.

File: cirq-core/cirq/optimizers/clifford_decomposition.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+    Args:
+        qubits: The list of qubits being operated on.
+        clifford_tableau: The Clifford Tableau for decomposition.
+
+    Returns:
+        A list of operations reconstructs the same Clifford tableau.
+
+    Raises:
+        ValueError: The length of input qubit mismatch with the size of tabluea.
```
Comment (MichaelBroughton) [inline cirq-core/cirq/optimizers/clifford_decomposition.py:103]: Nit: tabluea

File: cirq-core/cirq/optimizers/clifford_decomposition.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+):
+    protocols.act_on(ops.SWAP, args, qubits=[qubits[q1], qubits[q2]], allow_decompose=False)
+    operations.append(ops.SWAP(qubits[q1], qubits[q2]))
+
+
+def decompose_clifford_tableau_to_operations(
+    qubits: List['cirq.Qid'], clifford_tableau: qis.CliffordTableau
+) -> List[ops.Operation]:
+    """Decompose an n-qubit Clifford Tableau into a list of one/two qubit operations.
+
```
Comment (MichaelBroughton) [inline cirq-core/cirq/optimizers/clifford_decomposition.py:94]: Can you link out the paper in the docstring so if people look for this on the website it gets picked up ?

**Review by MichaelBroughton — APPROVED** [review]
LGTM. @Strilanc for visibility.

File: cirq-core/cirq/optimizers/clifford_decomposition.py (hunk 7 above)
Comment (ybc1991) [reply cirq-core/cirq/optimizers/clifford_decomposition.py]: Done

File: cirq-core/cirq/optimizers/clifford_decomposition.py (hunk 8 above)
Comment (ybc1991) [reply cirq-core/cirq/optimizers/clifford_decomposition.py:94]: Sure, added.

**Pull request merged**